Remove Block

jialuw96 opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jialuw96 commented:
@adowling2 I still cannot add reviewers.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -37,8 +37,10 @@
-from pyomo.contrib.doe.result import FisherResults, GridSearchResult
+#from pyomo.contrib.doe.scenario import Scenario_generator
+#from pyomo.contrib.doe.result import FisherResults, GridSearchResult
+from scenario import Scenario_generator
```
Why is this missing `pyomo.contrib.doe` at the beginning of the import statement? Happy to discuss in detail on Friday.

adowling2 replied:
We should change the name of `Scenario_generator` to `ScenarioGenerator` to be consistent with Python class naming conventions.

adowling2 replied:
Please check that the source file (.pptx?) used to make the UML diagram in the documentation has been added to GitHub. We may need to update the UML diagram as we make changes to the classes.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -396,12 +399,6 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             raise ValueError(self.mode+' is not a valid mode. Choose from "sequential_finite" and "direct_kaug".')
 
     def _sequential_finite(self, read_output, extract_single_model, store_output):
```
This function is missing a doc string.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -412,240 +409,99 @@ def _sequential_finite(self, read_output, extract_single_model, store_output):
+            #mod.T = pyo.Var(mod.t_con, bounds=[300, 700], within=pyo.NonNegativeReals)
 
-                output_record[no_s] = output_iter
+            mod.add_component('t0', pyo.Set(initialize=[0]))
```
Please update per our conversation.

jialuw96 replied:
Set up the model to have 3 options: `parmest`, `global`, `block`.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -412,240 +409,99 @@ def _sequential_finite(self, read_output, extract_single_model, store_output):
-                pickle.dump(output_record, f)
-                f.close()
+            def block_build(b,s):
+                self.create_model(m=b)
```
For `create_model` we should make the first argument `model=` to be more explicit that this is an existing model.

adowling2 replied:
Or `mod=`

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -412,240 +409,99 @@ def _sequential_finite(self, read_output, extract_single_model, store_output):
-                    mod_name = store_output + str(no_s) + '.csv'
-                    dataframe = extract_single_model(mod, square_result)
-                    dataframe.to_csv(mod_name)
+            mod = pyo.ConcreteModel()
```
Add a comment explaining what this is doing.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -412,240 +409,99 @@ def _sequential_finite(self, read_output, extract_single_model, store_output):
-
-            # return all models formed
-            self.models = models
+            mod.lsb = pyo.Block(mod.scena, rule=block_build)
```
What does `lsb` stand for? Can we think of a more descriptive variable name? If not, please add a comment explaining what this is doing (and define 'lsb').

jialuw96 replied:
I change `lsb` to `block` now.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -412,240 +409,99 @@ def _sequential_finite(self, read_output, extract_single_model, store_output):
 
-        FIM_analysis = FisherResults(list(self.param.keys()), self.measure, jacobian_info=None, all_jacobian_info=jac,
-                                    prior_FIM=prior_in_use, store_FIM=self.FIM_store_name, scale_constant_value=self.scale_constant_value)
+            # force all design variables in blocks be the same as global design variables
```
This will get updated per our conversation on how to generalize.

adowling2 commented on pyomo/contrib/doe/example/reactor_kinetics.py:
```diff
@@ -29,44 +29,45 @@
 from pyomo.dae import ContinuousSet, DerivativeVar
 import numpy as np
-from pyomo.contrib.doe import Measurements
+from measurements import Measurements
```
Why did this import statement change? We can discuss on Friday.

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -26,15 +26,15 @@
-    pandas as pd, pandas_available,
-    matplotlib as plt, matplotlib_available,
-)
+#from pyomo.common.dependencies import (
```
Why was this commented out?

jialuw96 replied:
We are not using these dependencies in `measurement`. Removed this sentence

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -26,15 +26,15 @@
 
 class Measurements:
-    def __init__(self, measurement_index_time, variance=None, ind_string='_index_'):
+    def __init__(self, self_define_res=None, measurement_index_time=None, variance=None, ind_string='_index_'):
```
After making the changes we discussed, update the doc string to include any new input arguments.

adowling2 reviewed (commented):
@jialuw96 Here are some comments. The import statements look strange. We can talk about this on Friday.

Once the interface is changed and the examples are updated, we need to write unit tests for new features. For instance, we are changing the interface to the Measurement class. We should write unit tests for that.

jialuw96 commented:
@adowling2 The compute_fim() function is ready for review, both sequential_finite mode and direct_kaug mode. `scenario` and `measurement` class and the reactor kinetics example are also updated per discussion. Thank you!

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -715,26 +540,25 @@ def _direct_kaug(self):
-                    dsdp_extract.append(zero_sens)
-            else:
-                kaug_no = col.index(measurement_name)
+        for mname in self.measure_name:
```
`for i, mname in enumerate(self.measure_name)`? Or do I not understand this code? What is going on with i?

jialuw96 replied:
I used these to change `C["CA", t]` to `C[CA, t]`, removing the "" outside a string. Now that we use the new way of generating measurements, this code is no longer needed. Removed.

adowling2 commented on pyomo/contrib/doe/example/reactor_kinetics.py:
```diff
@@ -29,209 +29,199 @@
-        m.index1 = False
-    elif model_form == 'dae-index-1':
-        m.index1 = True 
+    if model_option == "parmest":
```
Should this be...
if model_option == "parmest":
...
elif model_option == "global":
else:
   if not mod:
     raise error here?

jialuw96 replied:
This depends on if we want to create the model by ourself, or let user create model in global mode. Either is ok.

adowling2 replied:
I think we want to do some error checking to ensure the options passed to `model_option` are valid.

adowling2 replied:
We do need to test these value errors are being thrown.

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -61,216 +63,74 @@ def __init__(self, measurement_index_time, variance=None, ind_string='_index_'):
-        self.flatten_measure_timeset = flatten_measure_timeset
-
-    def _model_measure_name(self):
+        if self_define_res:
```
This is not the interface we discussed. We worked out having an `add_measurement` function that the user would call for each measurement and specify indices. This function would then build up the data stored in `self.measurement_name`.

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -61,216 +63,74 @@ def __init__(self, measurement_index_time, variance=None, ind_string='_index_'):
+            raise AttributeError("No measurement names provided. Define either self_define_res or measurement_var.")
+
+    def _measure_name(self):
         """Return pyomo string name
```
This doc string needs more explanation. The structure of self.measure_varname and other data are not clear.

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -44,14 +39,21 @@ def __init__(self, measurement_index_time, variance=None, ind_string='_index_'):
-                    values are a list of measuring time point.
-            For e.g., for the kinetics illustrative example, it should be {'C':{'CA':[0,1,..], 'CB':[0,2,...]}, 'k':[0,4,..]},
-            so the measurements are C[scenario, 'CA', 0]..., k[scenario, 0]....
+        self_define_res: a ``list`` of ``string``, containing the measurement variable names with indexs, 
```
This should be split into two functions instead of having all of these optional arguments. One function to directly specify the list of all measurement variable names. The section function to build up this list. That will make the documentation and examples much clearer. It will also be easier to test.

adowling2 commented on pyomo/contrib/doe/result.py:
```diff
@@ -85,7 +85,7 @@ def __init__(self, para_name, measure_object, jacobian_info=None, all_jacobian_i
 
 
-    def calculate_FIM(self, dv_values, result=None):
+    def calculate_FIM(self, dv_values, black_box_res=True, result=None):
```
Doc string needs to be updated to explain `black_box_res` option.

adowling2 reviewed (commented):
@jialuw96 See my comments.

jialuw96 commented:
@adowling2 Optimization function is ready for review. The discussed class is in specialSet.py. Thank you!

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -55,9 +57,8 @@ def __init__(self, param_init, design_variable_timepoints, measurement_object, c
-            A ``dictionary`` where keys are design variable names, values are its control time points.
-            If this design var is independent of time (constant), set the time to [0]
+        design_names:
+            A ``list`` of design variable names as string 
```
Does this doc string need to get updated? Is this now using SpecialSet?

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,37 +262,27 @@ def _optimize_stochastic_program(self, m):
         design_values:
-            a ``dict`` where keys are design variable names, 
-            values are a dict whose keys are time point and values are the design variable value at that time point
+            a ``list`` of design variable values 
```
Update doc string

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -412,66 +361,107 @@ def _sequential_finite(self, read_output, extract_single_model, store_output):
+            #def fix_design2(m,s,t):
+            #    return m.block[s].T[t] == m.T[t]
+
+            for name in self.design_name:
```
Does this work if the design variable is indexed in time? Is self.design_name a list of something else?

jialuw96 replied:
yes.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -695,15 +498,12 @@ def _direct_kaug(self):
-            var_name.append(name+'[0]')
-            var_dict[name+'[0]'] = self.param[name]
+            var_name.append(name)
+            var_dict[name] = self.param[name]
```
Is this copying self.param into var_dict? Likewise, is `var_name = list(self.param.keys())` equivalent?

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -715,26 +515,20 @@ def _direct_kaug(self):
         measurement_index = []
         # produce the sensitivity for fixed variables
         zero_sens = np.zeros(len(self.param))
+        print(col)
```
Remove this or add a more descriptive print statement.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -991,179 +776,140 @@ def _create_doe_model(self, no_obj=True):
+        # dictionary, keys are parameter name, values are the perturbation step 
+        self.eps_abs = scena_gen["eps-abs"]
+
+        # Create a global model 
```
This looks like duplicate code to earlier in the file. What is the difference?

jialuw96 replied:
make a private method here

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -991,179 +776,140 @@ def _create_doe_model(self, no_obj=True):
-                return 0
-
-        # initialize FIM
-        if self.fim_initial:
```
Is this removing the ability to initialize the FIM with a non-identity matrix, e.g., computed with k-aug? Isn't this an important feature?

jialuw96 replied:
added back.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -991,179 +776,140 @@ def _create_doe_model(self, no_obj=True):
+            q: parameter
             """
-            return m.FIM[j,d] == sum(sum(m.jac[z,j,i]*m.jac[z,d,i] for z in m.y_set) for i in m.tmea_set) + m.refele[j, d]*self.fim_scale_constant_value
+            return m.fim[p,q] == sum(m.jac[p,n]*m.jac[q,n] for n in mod.res) + m.priorFIM[p,q]*self.fim_scale_constant_value 
```
What about including the covariance structure for the measurements?

adowling2 replied:
We still need to discuss this.

adowling2 commented on pyomo/contrib/doe/example/reactor_kinetics.py:
```diff
@@ -29,209 +29,199 @@
-        for t in m.t:
-            m.dCdt_rule[z,'CC',t].deactivate()
+    if block:
+        for s in range(8):
```
Why is this eight hardcoded?

adowling2 replied:
Still need to update.

adowling2 commented on pyomo/contrib/doe/example/reactor_kinetics.py:
```diff
@@ -29,209 +29,199 @@
-    m.k2_pert_rule = pyo.Constraint(m.scena, m.t, rule=cal_kp2)
-    m.dCdt_rule = pyo.Constraint(m.scena,m.C_set, m.t, rule=dCdt_control)
+        # Control time points
+        mod.t_con = pyo.Set(initialize=t_control)
```
This seems redundant with the earlier code defining mod.t_con. Can this be streamlined?

jialuw96 replied:
streamlined by moving some common codes to the global part

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -25,252 +25,48 @@
 #  ___________________________________________________________________________
 
+import warnings
+from specialSet import SpecialSet
```
For simplicity, move SpecialSet into this file.

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -25,252 +25,48 @@
+        self.measurement_name = super().specify(self_define_res)
 
-    def _name_and_index_generator(self, all_info):
+    def add_elements(self, var_name, extra_index=None, time_index=[0]):
```
Why is add_elements defined here? Why not just inherit the version from SpecialSet?

jialuw96 replied:
using super()

adowling2 reviewed (commented):
@jialuw96 See my comments below. Nice work!

jialuw96 commented:
@adowling2 The grid search functions and variances are ready for review. Thank you!

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1241,28 +934,18 @@ def _fix_design(self, m, design_val, fix_opt=True, optimize_option=None):
+            var = cuid.find_component_on(m)
+            if fix_opt:
+                #getattr(m, dname)[time].fix(fix_v)    
+                var.fix(design_val[name])
```
Let's talk about using the DesignVariable class to store the nominal point and bounds. Would this be helpful for a sensitivity analysis?

adowling2 reviewed (commented):
Please are some comments to discuss today.

jialuw96 commented:
- [ ] variance for optimization 
- [ ] jacobian subset 
- [ ] Tictoctoker timer
- [ ] try enum https://docs.python.org/3/library/enum.html
- [ ] design variable class

jialuw96 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -482,228 +414,36 @@ def _sequential_finite(self, read_output, extract_single_model, store_output):
-                    measurement_accurate_time[j][no_t] = t_all[t_all.index(tt)]
-
         # set ub and lb to parameters
         for par in list(self.param.keys()):
```
Removed all list(dict.keys()) structure where it works without list()

adowling2 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -25,252 +25,273 @@
+                if it is a nested list, it is a ``list`` of ``lists``, they are different time set for different var in var_name
+        """
+        self._check_names(var_name, extra_index, time_index)
+        self.measurement_name =  super().add_elements(var_name=var_name, extra_index=extra_index, time_index=time_index)
```
Why not just call this "names"? That would make it easier to generalize across Measurements and DesignVariables.

adowling2 commented on pyomo/contrib/doe/specialSet.py:
```diff
@@ -0,0 +1,95 @@
+#  ___________________________________________________________________________
+import itertools
+
+class SpecialSet: 
```
Move this into Measurements.py and delete this file.

adowling2 reviewed (commented):
Here are my comments. Next time this is ready to review, tag me in a new post. When you edit an existing comment, I do not always get an update.

jialuw96 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,37 +271,27 @@ def _optimize_stochastic_program(self, m):
                     store_output = None, read_output=None, extract_single_model=None,
-                    formula='central', step=0.001,
-                    objective_option='det'):
+                    formula='central', step=0.001):
```
Ask user to import mode and formula class from Pyomo.doe, and give mode.central ~ here.

jialuw96 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -767,12 +506,62 @@ def _direct_kaug(self):
+            mod = self.discretize_model(mod)
+
+        # force design variables in blocks to be equal to global design values
+        for name in self.design_name:
```
It works since design_name is already a `specialSet`.

jialuw96 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -25,252 +25,273 @@
+            if nam not in self.measurement_name:
+                print("Measurement not in the set: ", nam)
 
+    def _generate_variance(self):
```
change to identity for identity function.

jialuw96 commented:
- [ ] Enum interface update 
- [ ] Update tutorial notebook, documentation
- [ ] Update tests

jsiirola committed:
Remove unreachable code, add missing call to value()

jialuw96 committed:
changes before switching branches

emma58 committed:
Finishing the paraboloid tests for the reduced inner repn transformation

jsiirola committed:
Remove unnecessary override

jsiirola committed:
Fix calculation of polynomial_degree(f(.)**0)

jsiirola committed:
Add optional optimizations for 0 in expression trees

...and standardize the order of argument processing, plus remove the
automatic optimization of f(.)**0 -> 1.

jsiirola committed:
Add tests for the 'zero' expression optimizations

jsiirola committed:
Check that all native_numeric_types are hashable

jsiirola committed:
NFC: improve docs

jsiirola committed:
NFC: fix typo, remove commented code

jsiirola committed:
Improve performance of LinearExpression.__init__

jsiirola committed:
Switch AMPLRepn.linear from list to dict

jsiirola committed:
Performance: faster printing of defined variables

jsiirola committed:
fix iteration over dict generators

jsiirola committed:
Fix typo

jsiirola committed:
Improve diff reporting (report actual baseline filename)

jsiirola committed:
Add the option to build againse pyomo-model-libraries branch

jsiirola committed:
Fix checkout of pyomo-model-libraries branch

jsiirola committed:
Reduce the size of the test problem slightly

jsiirola committed:
Merge pull request #2766 from emma58/pw-linear-or

contrib.piecewise: Adding transformation to outer representation GDP

jsiirola committed:
Fix for checking out pyomo-model-libraries

jsiirola committed:
Merge branch 'main' into nl-linear-dict

jsiirola committed:
Merge branch 'main' into pw-linear-ir-reduced

jsiirola committed:
Another attempt at identifying branch name

jsiirola committed:
Merge remote-tracking branch 'refs/remotes/me/nl-linear-dict' into nl-linear-dict

jsiirola committed:
Get the head (not base) ref

jsiirola committed:
Fall back on github.ref when not a PR

emma58 committed:
Merge pull request #2768 from emma58/pw-linear-ir-reduced

contrib.piecewise: Adding transformation to "reduced" inner representation GDP

jsiirola committed:
Merge pull request #2769 from jsiirola/nl-linear-dict

NL writer: improve linear datastructure

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
Merge branch 'main' into cyipopt-subclass

Robbybp committed:
call close before removing log file to attempt to remove lock

Robbybp committed:
add call to cyipopt.Problem.close at end of PyomoCyIpoptSolver.solve

Robbybp committed:
add a cyipopt.Problem.close call to a test that was missing it

jialuw96 committed:
debug Enum

jialuw96 committed:
update documentation first part; debugged measurement; create test_fim_doe new version

jialuw96 committed:
update test_reactor_example

jialuw96 committed:
update test_examples

jialuw96 committed:
debug tests. test_reactor_example pending debugging

jialuw96 committed:
debug all tests

jialuw96 committed:
update the rest of documentation

jialuw96 commented:
@adowling2 This is ready for review. Thank you!

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -150,8 +149,7 @@ def stochastic_program(self,  design_values, if_optimize=True, objective_option=
         design_values:
-            a ``dict`` where keys are design variable names, values are a dict whose keys are time point
-            and values are the design variable value at that time point
+            designVariable object
```
Is passing in design_values here needed? Aren't the design_vars (an instance of the DesignVariables class) already passed into the function? What is the purpose of passing this in again? Let's discuss.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -189,10 +187,9 @@ def stochastic_program(self,  design_values, if_optimize=True, objective_option=
         # store inputs in object
-        self.design_values = design_values
+        self.design_values = design_object.special_set_value
+        self.design_object = design_object
```
What information in the original instance of `design_object` was used to initialize the DesignOfExperiments class? What happens if a completely different `design_object` is passed in here? This interface design seems fragile. At a minimum, a lot of error checking is needed.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,37 +272,27 @@ def _optimize_stochastic_program(self, m):
+                    formula='central', step=0.001):
         """
         This function solves a square Pyomo model with fixed design variables to compute the FIM.
         It calculates FIM with sensitivity information from four modes:
```
Two modes?

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,37 +272,27 @@ def _optimize_stochastic_program(self, m):
 
         Parameters
         -----------
         design_values:
```
Need to update. See above comments.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,37 +272,27 @@ def _optimize_stochastic_program(self, m):
-            values are a dict whose keys are time point and values are the design variable value at that time point
+            designVariable object 
         mode:
             use mode='sequential_finite', 'sequential_sipopt', 'sequential_kaug', 'direct_kaug'
```
This list does not match the doc string above. Either fix this here or elaborate on the doc string.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -767,12 +507,62 @@ def _direct_kaug(self):
         return FIM_analysis
 
+    def _create_block(self):
+
```
This function needs a complete doc string

adowling2 commented on pyomo/contrib/doe/example/reactor_compute_FIM.py:
```diff
@@ -30,85 +30,57 @@
+
+    design_gen = DesignVariables()
+    design_gen.add_elements(total_name, time_index = dtime_index, values=exp1)
+    design_gen.add_bounds(upper_bound=upper_bound, lower_bound=lower_bound)
```
We need tests for these functions to ensure that if `upper_bounds` has the wrong length, the correct exception is thrown.

adowling2 reviewed (commented):
See comments below. This is getting really close to asking for the Pyomo team to review it.

mrmundt committed:
Reapply black to previously completed directories

jialuw96 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -25,252 +25,265 @@
-    pandas as pd, pandas_available,
-    matplotlib as plt, matplotlib_available,
-)
+class SpecialSet: 
```
also test special set

jialuw96 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -25,252 +25,265 @@
-        legal_t: if the time point is legal for this measurement 
-        string_name: one measurement pyomo string name for sequential 
+        if extra_index and len(extra_index) != num_var: 
+            warnings.warn("Extra_index is of different length with var_name. This warning indicates a potential modeling error.")
```
throw exception here

mrmundt committed:
Update dReadtheDocs config to address deprecation

mrmundt committed:
Merge pull request #2780 from mrmundt/readthedocs-deprs

[Docs] Update ReadtheDocs Configuration

mrmundt committed:
Merge branch 'main' into black/reapply

mrmundt committed:
Merge pull request #2775 from mrmundt/black/reapply

Reapply `black` to previously completed directories

jsiirola committed:
Remove the use of weakrefs in SymbolMap

michaelbynum committed:
unify appsi solver logging

michaelbynum committed:
Merge branch 'main' into appsi_logging

mrmundt committed:
Apply `black` to `pyomo/common` directory `py` files (#2776)

* Apply black to pyomo/common directory py files

* Accidentally removed a necessary space

* Restoring pyomo/core as they are not supposed to be in this change

* Fix missed string merges

* Fix odd indentation issue

mrmundt committed:
Apply `black` to `kernel` and `gdp` directory `py` files (#2784)

* Apply black to kernel and gdp directory py files

* Remove unnecessary comment

mrmundt committed:
Apply `black` to `pyomo/dataportal` directory `py` files (#2778)

* Apply black to pyomo/dataportal directory py files

* Apply odd mixed string fix

Robbybp committed:
fix error message

mrmundt committed:
Apply black to duality, mpec, and environ directory py files (#2779)

Robbybp committed:
test case with indexed block

mrmundt committed:
Apply black to pyomo/opt directory py files (#2782)

Robbybp committed:
make sure pyomo_nlp is available before checking isinstance(, PyomoNLP)

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into incidence-indexed-block

michaelbynum committed:
Merge pull request #2787 from michaelbynum/appsi_logging

Appsi logging

jsiirola committed:
Merge branch 'main' into symbolmap-no-weakref

jialuw96 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -150,8 +149,7 @@ def stochastic_program(self,  design_values, if_optimize=True, objective_option=
         Parameters
         -----------
         design_values:
```
This is changed to `design_vars` and `measurement_vars`

jialuw96 committed:
before changing k_aug

jialuw96 commented on pyomo/contrib/doe/measurements.py:
```diff
@@ -25,252 +25,265 @@
+        upper_bound: a ``list`` of upper bounds 
+        lower_bound: a ``list `` of lower bounds
+        """
+        if upper_bound:
```
It is moved to add_elements. `Upper_bounds` and `lower_bounds` now allows 3 kinds of inputs: if a list, they should be either a scalar number (will be broadcast to be the same length as names), a list (same length as names). I am confused what kind of dictionary you mentioned: the keys should be the string names, but user wouldn't know the string names at that time because this function is generating these names. And we need to check if the keys are the same as the generated names. Wouldn't a list be enough?

jialuw96 committed:
update measurements. Pending tests and tests for exceptions.

jsiirola committed:
Merge branch 'main' into symbolmap-no-weakref

jsiirola committed:
NLv2: Bugfix for shared named expressions

jialuw96 committed:
pass all current tests. Pending add new tests

jialuw96 committed:
add tests for model options, prior fim, add_element

jialuw96 committed:
pass all tests

jialuw96 commented:
@adowling2 This is ready for review. Thank you!

jsiirola committed:
Merge pull request #2789 from Robbybp/incidence-indexed-block

Test indexed block in Incidence Analysis

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -54,19 +75,17 @@ def __init__(self, param_init, design_variable_timepoints, measurement_object, c
+            A DesignVariable ``object`` which contains the Pyomo variable names and their corresponding indices
+            and bounds for experiment design of freedom 
+        measurement_vars:
+            A measurement ``object`` which contains the Pyomo variable names and their corresponding indices and 
```
measurement --> Measurement

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -177,22 +175,18 @@ def stochastic_program(self,  design_values, if_optimize=True, objective_option=
             a matrix used to initialize FIM matrix
         formula:
-            choose from 'central', 'forward', 'backward', None. This option is only used for 'sequential_finite' mode.
+            choose from 'central', 'forward', 'backward', None in the finite_difference_lib class
```
"Choose from finite_difference_lib.central, .forward, or .backward"

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,39 +264,27 @@ def _optimize_stochastic_program(self, m):
-
-        "Simultaneous_finite" mode is not included in this function.
+        It calculates FIM with sensitivity information from two modes:
+            1.  sequential_finite: use finite difference scheme to evaluate sensitivity
```
"calculation_mode.sequential_finite: sequentially solve square problems and use finite difference approximation"

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,39 +264,27 @@ def _optimize_stochastic_program(self, m):
-        "Simultaneous_finite" mode is not included in this function.
+        It calculates FIM with sensitivity information from two modes:
+            1.  sequential_finite: use finite difference scheme to evaluate sensitivity
+            2.  direct_kaug: use k_aug to evaluate sensitivity
```
"calculation_mode.direct_kaug: solve a single square problem then extract derivatives using NLP sensitivity theory"

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,39 +264,27 @@ def _optimize_stochastic_program(self, m):
-            values are a dict whose keys are time point and values are the design variable value at that time point
         mode:
-            use mode='sequential_finite', 'sequential_sipopt', 'sequential_kaug', 'direct_kaug'
+            use sequential_finite, direct_kaug in calculation_mode.
```
"use calculation_mode.sequential_finite or calculation_mode.direct_kaug"

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -349,29 +304,27 @@ def compute_FIM(self, design_values, mode='sequential_finite', FIM_store_name=No
             scenario index is the number of the scenario outputs which is stored.
         formula:
-            choose from 'central', 'forward', 'backward', None. This option is only used for 'sequential_finite' mode.
+            choose from 'central', 'forward', 'backward' in finite_difference_lib, None. This option is only used for sequential_finite mode.
```
See same comment as above.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -297,39 +264,27 @@ def _optimize_stochastic_program(self, m):
                     store_output = None, read_output=None, extract_single_model=None,
-                    formula='central', step=0.001,
-                    objective_option='det'):
+                    formula='central', step=0.001):
```
Why is the formula not an Enum?

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -811,51 +589,44 @@ def _finite_calculation(self, output_record, scena_gen):
-    def run_grid_search(self, design_values, design_ranges, design_dimension_names, 
-                    design_control_time, mode='sequential_finite', tee_option=False, 
+    def run_grid_search(self, design_ranges, design_dimension_names, 
+                     mode='sequential_finite', tee_option=False, 
```
Why is the default mode not an Enum?

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -811,51 +589,44 @@ def _finite_calculation(self, output_record, scena_gen):
-            a ``list`` of control time points that should be fixed to the values in dv_ranges
         mode:
-            use mode='sequential_finite', 'sequential_sipopt', 'sequential_kaug', 'direct_kaug'
+            use mode='sequential_finite', 'direct_kaug'
```
Update this doc string to use Enum. See comments above for suggestions/formatting.

adowling2 reviewed (commented):
Here are some minor comments. I think we are ready to open a PR into Pyomo.

jsiirola committed:
Add error checking to addSymbol

jsiirola committed:
Fix inconsistencies in symbol map interface

jsiirola committed:
Removing an unnecessary test (that will always be true)

jsiirola committed:
Adding additional SymbolMap testing

jsiirola committed:
Reverting 4071f2f

jsiirola committed:
NFC: Updating comments

jsiirola committed:
Merge pull request #2790 from jsiirola/nl-shared-expr-fix

NLv2: Bugfix for shared named expressions

mrmundt committed:
Set up git-blame-ignore file

jsiirola committed:
Merge pull request #2791 from jsiirola/symbolmap-no-weakref

SymbolMap: remove the use of weakrefs

mrmundt committed:
Add black linting action

mrmundt committed:
Adjust contrib docs for black/spell check expectations

mrmundt committed:
Add false typos and ignore a file type

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

Robbybp committed:
call close() on user-created problem object rather than private _problem attribute

Robbybp committed:
note that close must be called to force Ipopt to release its lock on the log file

Commit:
Fix community.rst

Import repeated

mrmundt committed:
Merge pull request #2792 from DavidDB33/patch-1

Fix community.rst

jialuw96 committed:
clean up docs

jialuw96 commented:
@adowling2 The repo is cleaned up and ready for review. Thank you.

adowling2 commented:
In parallel to me reviewing this here, let’s open a PR for your branch into the main Pyomo repo.

jialuw96 commented:
@adowling2 The PR is opened in https://github.com/jialuw96/pyomo/tree/Feb_block_doe.

mrmundt committed:
Replace pandas append with concat

mrmundt committed:
Change concat usage for pandas 2.0.0

mrmundt committed:
Merge pull request #2795 from mrmundt/pandas-2.0.0

Update to Parmest for pandas 2.0.0 release

mrmundt committed:
Fix typos in all dirs to which black has been applied already

mrmundt committed:
Apply `black` to `pyomo/contrib` directory `py` files (#2781)

* Apply black to pyomo/contrib directory py files

* Apply black to residual pyomo/contrib files

* Clean up mixed up branch

* Apply fixes to odd comment corrections

* Address one missed awkward indentation

mrmundt committed:
Add SHA for black/contrib

mrmundt committed:
Merge branch 'main' into auto-linting

mrmundt committed:
Merging main

mrmundt committed:
Apply typo fixes to contrib

mrmundt committed:
Reapply black to some dual modified files

mrmundt committed:
Temporarily turn off linting to verify changes

mrmundt committed:
Manually correct missed typos

emma58 committed:
Filtering out degenerate simplices after Delaunay triangulation, raising a DeveloperError when it appears that failed, warning user about points dropped in the triangulation

emma58 committed:
Blackification

mrmundt committed:
Apply `black` to `dae` and `network` directory `py` files (#2783)

* Apply black to dae and network directory py files

* NFC: cleaning up comment formatting

---------

Co-authored-by: John Siirola <jsiirola@users.noreply.github.com>

mrmundt committed:
Merge branch 'main' into auto-linting

mrmundt committed:
Apply typo correctionss post DAE/Network black merge

mrmundt committed:
Apply `black` to `pyomo/core` directory `py` files (#2777)

* Apply black to pyomo/core directory py files

* NFC: cleaning up some comment/string formatting

---------

Co-authored-by: John Siirola <jsiirola@users.noreply.github.com>

mrmundt committed:
Merge remote-tracking branch 'origin' into auto-linting

mrmundt committed:
Apply typo fixes and last pass at black formatting

mrmundt committed:
Turn black linting on for PRs

mrmundt committed:
Change black arguments

mrmundt committed:
Apply black to residual examples aftering converting to Python3

mrmundt committed:
Look at file directory

mrmundt committed:
Lock black to v22.6

emma58 committed:
Adding tests for degenerate simplices and redundant points

emma58 committed:
Fixing a typo

emma58 committed:
Adding argument for tabular data: a dictionary mapping points to nonlinear function values

emma58 committed:
NFC: applying black

emma58 committed:
Removing a redundant line

emma58 committed:
NFC: updating comments

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

conflicts were resolved by taking the cyipopt-subclass files and running
black on them.

Robbybp committed:
run black on files I missed

Robbybp committed:
re-run black, now with --skip-string-normalization

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

emma58 committed:
Sorting the simplices since apparently Delaunay behaves differently on Windows

emma58 committed:
Sorting the simplex indices when we store them in the pw-linear function, which means that later sorting the simplices means something.

jsiirola committed:
Update magic constant for visitor tests for python 3.11.2

emma58 committed:
Changing tests to just verify that there are 6 full-dimensional simplices inside of each of the 8 cubes of the grid.

emma58 committed:
Merge branch 'main' into pw-linear-fix

mrmundt committed:
Merge pull request #2799 from jsiirola/fix-visitor-tests-3.11

Update magic constant for visitor tests for python 3.11.2

mrmundt committed:
Apply black 23.3.0 - adding white space

mrmundt committed:
Unpin black version

mrmundt committed:
Merge branch 'main' into auto-linting

mrmundt committed:
Add mass black reapplication to ignore list

mrmundt committed:
Update instructions for local configuration, not global

mrmundt committed:
Fix CHANGELOG line to reflect typo change

jsiirola committed:
Merge branch 'main' into pw-linear-fix

mrmundt committed:
Update contributer's guide with full black command

mrmundt committed:
Residual awkward line combinations from black

mrmundt committed:
Apply black change

mrmundt committed:
Apply comments per @jsiirola

mrmundt committed:
Standardization of word usage

mrmundt committed:
Merge pull request #2800 from mrmundt/auto-linting

Auto-Linting: Spelling Black Style Checker

emma58 committed:
It's not always my fault if the simplices are degenerate--keeping track of where to place the blame and throwing either a ValueError or DeveloperError accordingly.

emma58 committed:
Merge branch 'pw-linear-fix' of github.com:emma58/pyomo into pw-linear-fix

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
Addressing reviewer comments

jsiirola committed:
Applying black

jsiirola committed:
NFC: fixing spelling

emma58 committed:
Skipping test when numpy not available

emma58 committed:
Merge branch 'main' into pw-linear-fix

emma58 committed:
Merge branch 'main' into pw-linear-fix

emma58 committed:
Merge branch 'pw-linear-fix' of github.com:emma58/pyomo into pw-linear-fix

jsiirola committed:
Correct implementation of polynomial_degree() for **0

emma58 committed:
Changing logging to be in __name__

emma58 committed:
merging main

emma58 committed:
No typos here

jsiirola committed:
Updating tests for **0

jsiirola committed:
Merge pull request #2797 from emma58/pw-linear-fix

contrib.piecewise: Fixing a bug caused by degenerate simplices

emma58 committed:
Resolving merge conflict

emma58 committed:
Passing dimension through so that the we know it to filter the simplices from the Delaunay triangulation

jsiirola committed:
Improve formatting for DeveloperError, MouseTrap messages

jsiirola committed:
Switch signature for sympy _operatorMap dispatcher

jsiirola committed:
Switch Sympy2PyomoVisitor to use public sympy API

jsiirola committed:
Update Pyomo2SympyVisitor to resolve NPV expressions to constants

jsiirola committed:
Raise an exception when converting expressions with ExternalFunctions to sympy

jsiirola committed:
Switch cnf_walker to use the visitors from sympy_tools

jsiirola committed:
NFC: comment updates

jsiirola committed:
logical_to_linear: raise exception for logical constraints with relational expression nodes

jsiirola committed:
Adding tests

jsiirola committed:
Refining MouseTrap message

jsiirola committed:
Improve warning when dynamically registering numeric types

jsiirola committed:
Resolving test failures

jsiirola committed:
NFC: fix typos

jsiirola committed:
Merge pull request #2805 from jsiirola/improved-error-formatting

Improve formatting for DeveloperError, MouseTrap messages

jsiirola committed:
Merge branch 'main' into sympy-walker-updates

jsiirola committed:
Fix signature for external function callback

jsiirola committed:
Add logical_to_linear tests

jsiirola committed:
NFC: fix typo

jsiirola committed:
Merge pull request #2798 from emma58/pw-linear-funct-from-tabular-data-really

contrib.piecewise: Construct PiecewiseLinearFunction from tabular data

jsiirola committed:
Updating imports to reflect TPL from cython builds

jsiirola committed:
Merge pull request #2806 from jsiirola/sympy-walker-updates

Improve conversions between Pyomo and Sympy expressions

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

Robbybp committed:
update formatting for consistency with black

Robbybp committed:
fix indentation

Robbybp committed:
comma at end of arg list

Robbybp committed:
remove whitespace

jsiirola committed:
NLv2: resolve NPV constants in LinearExpression to native type

jsiirola committed:
Improve initialization of LinearExpression objects

jsiirola committed:
Remove portability code for python < 3.7

jsiirola committed:
Additional LinearExpression NLv2 test

jsiirola committed:
NPC: apply black

jsiirola committed:
Fix index math error

jsiirola committed:
Revert change in exception type from #2806

jsiirola committed:
Support fallback to numeric differentiation

jsiirola committed:
NFC: apply black

jialuw96 committed:
completing new API

jialuw96 committed:
address comments. Pending updates tests

jsiirola committed:
Merge pull request #2811 from jsiirola/fix-nlv2-npv-linear-const

NLv2: resolve NPV constants in LinearExpressions

jialuw96 committed:
debugged computeFIM and stochasticProgram

jialuw96 committed:
debugged. now pass all tests

Robbybp committed:
Merge branch 'main' of https://github.com/pyomo/pyomo into cyipopt-subclass

Robbybp committed:
add flag and check to catch if solve is calledwhen cyipopt.Problem.__init__ has not been called

blnicho committed:
NFC: Fixing string formatting

blnicho committed:
Merge pull request #2760 from Robbybp/cyipopt-subclass

Refactor CyIpopt interface to subclass `cyipopt.Problem`

jsiirola committed:
Merge branch 'main' into expr-multiple-dispatch

jsiirola committed:
NFC: Apply black

blnicho committed:
Merge pull request #2722 from jsiirola/expr-multiple-dispatch

Rework expression generation to leverage multiple dispatch

jsiirola committed:
GHA: link test is a prereq only for PR/main builds

jsiirola committed:
Fix GHA syntax

michaelbynum committed:
Merge pull request #2818 from jsiirola/gha-lint-branch

GHA: lint test is a prereq only for PR/main builds

jsiirola committed:
Merge branch 'main' into calc-var-value-diff-mode

jsiirola committed:
Adding tests for external functions

jsiirola committed:
Split external function test into 2 tests

jsiirola committed:
missing import

jsiirola committed:
Correct import location

jialuw96 committed:
Merge branch 'Feb_block_doe' of https://github.com/jialuw96/pyomo into Feb_block_doe

# Conflicts:
#	doc/OnlineDocs/contributed_packages/doe/doe.rst
#	pyomo/contrib/doe/__init__.py
#	pyomo/contrib/doe/doe.py
#	pyomo/contrib/doe/example/reactor_compute_FIM.py
#	pyomo/contrib/doe/example/reactor_grid_search.py
#	pyomo/contrib/doe/example/reactor_kinetics.py
#	pyomo/contrib/doe/example/reactor_optimize_doe.py
#	pyomo/contrib/doe/fim_doe_tutorial.ipynb
#	pyomo/contrib/doe/measurements.py
#	pyomo/contrib/doe/result.py
#	pyomo/contrib/doe/scenario.py
#	pyomo/contrib/doe/tests/test_example.py
#	pyomo/contrib/doe/tests/test_fim_doe.py
#	pyomo/contrib/doe/tests/test_reactor_example.py

jialuw96 committed:
restore files

jialuw96 committed:
ran black

jsiirola committed:
Merge pull request #2812 from jsiirola/calc-var-value-diff-mode

Improve robustness of `calculate_variable_from_constraint()`

jialuw96 committed:
changed scenario.py. passed all tests

jialuw96 committed:
changed names. passd all tests

jialuw96 committed:
added test. passed all test.

jialuw96 committed:
updated document

MLopez-Ibanez committed:
[DOC] Remove instructions for python <= 3.0

because Pyomo requires Python >= 3.7.

blnicho committed:
Adding blank line removed from doctest by mistake

mrmundt committed:
Merge pull request #2822 from MLopez-Ibanez/patch-1

[DOC] Remove instructions for python <= 3.0

jialuw96 committed:
addressing comments. pass tests

jialuw96 committed:
finish updating enum. passed all tests

jialuw96 committed:
addressing comments. passed all tests.

jialuw96 committed:
scenario dictionary changed to named tuple.

jialuw96 committed:
Merge branch 'main' into Feb_block_doe

jialuw96 committed:
update doc and notebook

jialuw96 committed:
run black

jialuw96 committed:
checked spells, fix some errors in documentation

jialuw96 committed:
delete non-used numpy import; try to import numpy from pyomo.common.dependencies

jialuw96 committed:
skip numpy if not numpy_available in tests

jialuw96 committed:
run black

jialuw96 committed:
update doe.rst according to comments

jialuw96 committed:
solve calculate_FIM() issue, rename to result_analysis, change the input

jialuw96 committed:
before try context manager

jialuw96 committed:
use context manager. run black

jialuw96 committed:
finish change fim_doe_tutorials

jialuw96 committed:
fix typo

blnicho committed:
NFC: Fixing Sphinx warnings and docstring formatting

blnicho committed:
NFC: Fix black formatting

blnicho committed:
NFC: Updates to doe.rst

adowling2 committed:
Removing unused comment

adowling2 committed:
Address model Pyomo.DoE review comments

mod in doc string

adowling2 committed:
Added missing m: Pyomo model from doc string

adowling2 committed:
Fix black formatting error?

adowling2 committed:
Fixed variable name typo
